### DISCUSSION TITLE
Custom interface icon like the standard fields. They use svg and a "|preview" property

### DISCUSSION THREAD

**Discussion by kevincam3** [directus/directus#16629] [Archive]
I tried this same as the built in interfaces but it doesn't work for me even though the file is at that location.

import PreviewSVG from './preview.svg?raw';
	preview: PreviewSVG,

I got this error.

Error: Could not load /home/kevincam/PhpstormProjects/directus/app/date-created/src/preview.svg?raw (imported by src/index.ts): ENOENT: no such file or directory, open '/home/kevincam/PhpstormProjects/directus/app/date-created/src/preview.svg?raw'

**Comment by br41nslug** [comment]
have you tried `import PreviewSVG from './preview.svg';`? If that fails because of import type (im not sure if the required plugins for importing svg are available out of the box). `preview` expects a string so you can always work around it by either using the SVG source directly here `preview: '<svg>....'` or turning the svg into a JS file and importing that like `preview.js` -> `module.exports = 'SVG source'`.

**Reply by kevincam3** [reply]
Thanks for that tip. Since I'm using typescript I just renamed it to preview.ts and then did

export default `<svg></svg>` using back ticks instead of quotes and it works just fine. Thanks again. I would love to know why the default interfaces don't have this problem and can import the svg without doing this.

**Reply by br41nslug** [reply]
Why this works for the default interfaces is the bundler is different between the core app which uses Vite (https://vitejs.dev/) and for extensions which use Rollup (https://rollupjs.org/guide/en/). While Rollup does support importing non JS/TS scripts it requires plugins for this that we do not provide out of the box 😬

**Comment by kevincam3** [comment]
But Vite uses Rollup under the hood

**Reply by br41nslug** [reply]
Does it? TIL :O But anyway the point was there are enough differences between them for them to behave differently (perhaps vite automatically enables these plugins where plain rollup does not)

**Reply by kevincam3** [reply]
But the core extension are part of the Vue app, so they also get handled by vite. Now that I think about it why are you saying they get handled by a different bundler? Are you talking about the API?

**Reply by br41nslug** [reply]
> But the core extension are part of the Vue app, so they also get handled by vite. Now that I think about it why are you saying they get handled by a different bundler?

Yes the difference is between core extensions and custom extensions. Because your custom extension first gets compiled with Rollup (`npm run build`) before loading it into the App and the problem you're describing is that the plain rollup build for custom extensions does not include the required plugins to import non-js files (like `svg`) so this results in the imported svg file not being present in the built extension output (or the build failing completely).

**Reply by kevincam3** [reply]
Ah, ok so then I need to install the plugin in the extensions package.json and not the vite.config file. I'll take a look at that option.

**Reply by kevincam3** [reply]
Or add a rollup config file and do it.

**Reply by br41nslug** [reply]
Extending the rollup config was discussed here https://github.com/directus/directus/discussions/6599 and implemented here https://github.com/directus/directus/pull/7332
You should be able to add plugins to the rollup config with this although i can't say i've personally used it before.

**Reply by kevincam3** [reply]
Excellent. I got it working now. Here's what I had to do incase someone else wants to use a custom svg image for their extension.

1. Installed this rollup plugin from your extension's directory so it's in your extensions package.json `npm i rollup-plugin-inline-svg -D`
2. Create your rollup config file in the same directory make sure to use this exact name "extension.config.js"
3. Put the following code inside the extension.config.js file to use the plugin we installed earlier

    ```js
    import InlineSvg from 'rollup-plugin-inline-svg';
    
    export default {
	    plugins: [InlineSvg()],
    };
    ```

4. To stop typescript from complaining about the import, if you're using typescript, then in the src folder of your extension there is a shims.d.ts file add the following line after what is already there:

    ```js
    declare module '*.svg';
    ```

5. last thing is to import and add the preview property to your index.ts file

    ```js
    import PreviewSVG from './preview.svg';
    
    export default defineInterface({
	    id: 'tcm-date-created',
	    name: 'Date Created',
	    icon: 'box',
	    description: 'Save the date the entry was created',
	    component: InterfaceComponent,
	    options: null,
	    types: ['string'],
	    group: 'other',
	    preview: 'PreviewSVG',
    });
    ```

NOTE: in my case the svg file preview.svg is in the src directory.

![image](https://user-images.githubusercontent.com/870934/204383859-b921ffec-7e18-4204-8dcc-0a2d416ecb30.png)
